File: ProjectPlan/generate_sprocs.js
const fs = require('fs');
const path = require('path');
const file = path.resolve(__dirname, 'Gayatri-Portal-DB-script.txt');
const text = fs.readFileSync(file, 'utf8');
const tables = [];
const tableRegex = /CREATE TABLE \[dbo\]\.\[([^\]]+)\]\(([\s\S]*?)\)\s*ON \[PRIMARY\]/g;
let tableMatch;
while ((tableMatch = tableRegex.exec(text)) !== null) {
  const name = tableMatch[1];
  const colsText = tableMatch[2];
  const cols = [];
  for (const rawLine of colsText.split(/\r?\n/)) {
    let line = rawLine.trim();
    if (!line || line.startsWith('PRIMARY KEY') || line.startsWith('CONSTRAINT') || /^\)\s*ON/.test(line)) continue;
    line = line.replace(/,$/, '');
    const colm = line.match(/^\[([^\]]+)\]\s+([^\s]+(?:\([^\)]*\))?)(.*)$/);
    if (!colm) continue;
    const colname = colm[1];
    const ctype = colm[2];
    const rest = colm[3];
    const normalizedType = ctype.toUpperCase();
    if (!/^(INT|BIGINT|SMALLINT|TINYINT|BIT|DECIMAL|MONEY|FLOAT|DATETIME2?|DATE|TIME|NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT)/.test(normalizedType)) continue;
    const isIdentity = rest.includes('IDENTITY');
    const isNullable = !rest.includes('NOT NULL');
    cols.push({ colname, ctype, isNullable, isIdentity, rest: rest.trim() });
  }
  let pk = null;
  const pkmatch = tableMatch[0].match(/PRIMARY KEY CLUSTERED \([^\)]*\[([^\]]+)\] ASC/);
  if (pkmatch) pk = pkmatch[1];
  else {
    const identityCol = cols.find(c => c.isIdentity);
    pk = identityCol ? identityCol.colname : (cols[0] ? cols[0].colname : null);
  }
  if (pk) tables.push({ name, pk, cols });
}
function paramType(coltype) {
  const t = coltype.toUpperCase();
  if (/^(NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT)/.test(t)) return t.includes('MAX') ? 'NVARCHAR(MAX)' : t;
  if (/^(DECIMAL|MONEY|FLOAT)/.test(t)) return t;
  if (/^(DATETIME2|DATETIME|DATE|TIME)/.test(t)) return t;
  if (/^(INT|SMALLINT|TINYINT|BIGINT)/.test(t)) return t;
  if (/^BIT/.test(t)) return 'BIT';
  return t;
}
const out = [];
out.push('USE [GayatriCateringPortal];\nGO\n\nSET ANSI_NULLS ON\nGO\nSET QUOTED_IDENTIFIER ON\nGO\n');
for (const { name, pk, cols } of tables) {
  const nonIdCols = cols.filter(c => !c.isIdentity);
  const params = nonIdCols.map(c => {
    const ptype = paramType(c.ctype);
    const defaultVal = c.isNullable && /^(NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT|DATETIME)/i.test(ptype) ? ' = NULL' : (c.isNullable && ptype === 'BIT' ? ' = NULL' : '');
    return `    @${c.colname} ${ptype}${defaultVal}`;
  });
  out.push(`-- Procedures for ${name}`);
  out.push(`IF OBJECT_ID(N'SP_Create${name}', N'P') IS NOT NULL DROP PROCEDURE SP_Create${name};`);
  out.push('GO');
  out.push(`CREATE PROCEDURE SP_Create${name}`);
  if (params.length) out.push(`(${params.join(',\n')})`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (params.length) {
    const colsList = nonIdCols.map(c => `[${c.colname}]`).join(', ');
    let valsList = nonIdCols.map(c => `@${c.colname}`).join(', ');
    if (nonIdCols.some(c => c.colname.toLowerCase() === 'createddate')) {
      valsList = valsList.replace('@CreatedDate', 'COALESCE(@CreatedDate, GETDATE())');
    }
    out.push(`    INSERT INTO [dbo].[${name}] (${colsList})`);
    out.push(`    VALUES (${valsList});`);
    out.push('    SELECT SCOPE_IDENTITY() AS NewId;');
  } else {
    out.push(`    INSERT INTO [dbo].[${name}] DEFAULT VALUES;`);
    out.push('    SELECT SCOPE_IDENTITY() AS NewId;');
  }
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'SP_Update${name}', N'P') IS NOT NULL DROP PROCEDURE SP_Update${name};`);
  out.push('GO');
  const updParams = [`    @${pk} ${paramType(cols.find(c => c.colname === pk).ctype)}`, ...params];
  out.push(`CREATE PROCEDURE SP_Update${name}`);
  out.push(`(${updParams.join(',\n')})`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (params.length) {
    const setList = nonIdCols.map(c => `    [${c.colname}] = @${c.colname}`).join(',\n');
    out.push(`    UPDATE [dbo].[${name}]`);
    out.push('    SET ' + setList);
    out.push(`    WHERE [${pk}] = @${pk};`);
  }
  out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'SP_Get${name}ById', N'P') IS NOT NULL DROP PROCEDURE SP_Get${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE SP_Get${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${paramType(cols.find(c => c.colname === pk).ctype)}`);
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  out.push(`    SELECT * FROM [dbo].[${name}] WHERE [${pk}] = @${pk};`);
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'Get${name}', N'P') IS NOT NULL DROP PROCEDURE Get${name};`);
  out.push('GO');
  out.push(`CREATE PROCEDURE Get${name}`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  const whereDeleted = cols.some(c => c.colname.toLowerCase() === 'isdeleted');
  out.push(`    SELECT * FROM [dbo].[${name}]${whereDeleted ? ' WHERE [IsDeleted] = 0' : ''};`);
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'Delete${name}ById', N'P') IS NOT NULL DROP PROCEDURE Delete${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE Delete${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${paramType(cols.find(c => c.colname === pk).ctype)}`);
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (whereDeleted) {
    out.push(`    UPDATE [dbo].[${name}] SET [IsDeleted] = 1 WHERE [${pk}] = @${pk};`);
  } else {
    out.push(`    DELETE FROM [dbo].[${name}] WHERE [${pk}] = @${pk};`);
  }
  out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'ActiveInActive${name}ById', N'P') IS NOT NULL DROP PROCEDURE ActiveInActive${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE ActiveInActive${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${paramType(cols.find(c => c.colname === pk).ctype)},`);
  out.push('    @IsActive BIT');
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (cols.some(c => c.colname.toLowerCase() === 'isactive')) {
    out.push(`    UPDATE [dbo].[${name}] SET [IsActive] = @IsActive WHERE [${pk}] = @${pk};`);
    out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  } else {
    out.push(`    RAISERROR('Table [${name}] does not contain an IsActive column.', 16, 1);`);
    out.push('    RETURN;');
  }
  out.push('END');
  out.push('GO');
}
const outputPath = path.resolve(__dirname, 'Gayatri-Portal-DB-sprocs.sql');
fs.writeFileSync(outputPath, out.join('\n'), 'utf8');
console.log('Wrote', outputPath, 'for', tables.length, 'tables.');

File: ProjectPlan/generate_sprocs.py
import re
from pathlib import Path

path = Path(__file__).parent / 'Gayatri-Portal-DB-script.txt'
text = path.read_text(encoding='utf-8')
blocks = re.split(r'GO\s*\r?\n', text)
tables = []
for block in blocks:
    m = re.search(r'CREATE TABLE \[dbo\]\.\[([^\]]+)\]\((.*?)\)\s*ON \[PRIMARY\]', block, re.S)
    if m:
        name = m.group(1)
        cols_text = m.group(2)
        cols = []
        for line in cols_text.splitlines():
            line = line.strip()
            if not line or line.startswith('PRIMARY KEY') or line.startswith('CONSTRAINT') or line.startswith(')'):
                continue
            line = line.rstrip(',')
            colm = re.match(r'\[([^\]]+)\]\s+([^\s]+(?:\([^\)]*\))?)(.*)$', line)
            if colm:
                colname = colm.group(1)
                ctype = colm.group(2)
                rest = colm.group(3)
                is_identity = 'IDENTITY' in rest
                is_nullable = 'NOT NULL' not in rest
                cols.append((colname, ctype, is_nullable, is_identity, rest.strip()))
        pk = None
        pkmatch = re.search(r'PRIMARY KEY CLUSTERED \([^\)]*\[([^\]]+)\] ASC', block)
        if pkmatch:
            pk = pkmatch.group(1)
        else:
            for cname, ctype, nullable, ident, rest in cols:
                if ident:
                    pk = cname
                    break
            if not pk and cols:
                pk = cols[0][0]
        tables.append((name, pk, cols, block))


def param_type(coltype):
    t = coltype.upper()
    if t.startswith(('NVARCHAR', 'VARCHAR', 'NCHAR', 'CHAR', 'TEXT')):
        return 'NVARCHAR(MAX)' if 'MAX' in t else t
    if t.startswith(('DECIMAL', 'MONEY', 'FLOAT')):
        return t
    if t.startswith(('DATETIME2', 'DATETIME', 'DATE', 'TIME')):
        return t
    if t.startswith(('INT', 'SMALLINT', 'TINYINT', 'BIGINT')):
        return t
    if t.startswith('BIT'):
        return 'BIT'
    return t

out = []
out.append('USE [GayatriCateringPortal];\nGO\n\nSET ANSI_NULLS ON\nGO\nSET QUOTED_IDENTIFIER ON\nGO\n')
for table, pk, cols, block in tables:
    non_id_cols = [c for c in cols if not c[3]]
    params = []
    for colname, ctype, nullable, ident, rest in non_id_cols:
        ptype = param_type(ctype)
        default = ' = NULL' if nullable else ''
        if ptype.upper() in ('BIT',) and nullable:
            default = ' = NULL'
        elif ptype.upper().startswith(('NVARCHAR', 'VARCHAR')):
            default = ' = NULL' if nullable else ''
        params.append(f'    @{colname} {ptype}{default}')
    out.append(f'-- Procedures for {table}\n')
    out.append(f'IF OBJECT_ID(N''SP_Create{table}'', N''P'') IS NOT NULL DROP PROCEDURE SP_Create{table};\nGO')
    out.append(f'CREATE PROCEDURE SP_Create{table}')
    if params:
        out.append('(' + '\n' + ',\n'.join(params) + '\n)')
    out.append('AS\nBEGIN\n    SET NOCOUNT ON;\n')
    if params:
        cols_list = ', '.join(f'[{c[0]}]' for c in non_id_cols)
        vals_list = ', '.join(f'@{c[0]}' for c in non_id_cols)
        if any(c[0].lower() == 'createddate' for c in non_id_cols):
            vals_list = vals_list.replace('@CreatedDate', 'COALESCE(@CreatedDate, GETDATE())')
        out.append(f'    INSERT INTO [dbo].[{table}] ({cols_list})\n    VALUES ({vals_list});\n    SELECT SCOPE_IDENTITY() AS NewId;\n')
    else:
        out.append(f'    INSERT INTO [dbo].[{table}] DEFAULT VALUES;\n    SELECT SCOPE_IDENTITY() AS NewId;\n')
    out.append('END\nGO\n')

    out.append(f'IF OBJECT_ID(N''SP_Update{table}'', N''P'') IS NOT NULL DROP PROCEDURE SP_Update{table};\nGO')
    upd_params = [f'    @{pk} {param_type(next(c[1] for c in cols if c[0] == pk))}']
    upd_params += params
    out.append(f'CREATE PROCEDURE SP_Update{table}')
    out.append('(' + '\n' + ',\n'.join(upd_params) + '\n)')
    out.append('AS\nBEGIN\n    SET NOCOUNT ON;\n')
    if params:
        set_list = ',\n        '.join(f'[{c[0]}] = @{c[0]}' for c in non_id_cols)
        out.append(f'    UPDATE [dbo].[{table}]\n    SET {set_list}\n    WHERE [{pk}] = @{pk};\n')
    out.append('    SELECT @@ROWCOUNT AS RowsAffected;\nEND\nGO\n')

    out.append(f'IF OBJECT_ID(N''SP_Get{table}ById'', N''P'') IS NOT NULL DROP PROCEDURE SP_Get{table}ById;\nGO')
    out.append(f'CREATE PROCEDURE SP_Get{table}ById\n(\n    @{pk} {param_type(next(c[1] for c in cols if c[0] == pk))}\n)\nAS\nBEGIN\n    SET NOCOUNT ON;\n    SELECT * FROM [dbo].[{table}] WHERE [{pk}] = @{pk};\nEND\nGO\n')

    out.append(f'IF OBJECT_ID(N''Get{table}'', N''P'') IS NOT NULL DROP PROCEDURE Get{table};\nGO')
    out.append(f'CREATE PROCEDURE Get{table}\nAS\nBEGIN\n    SET NOCOUNT ON;\n    SELECT * FROM [dbo].[{table}]')
    if any(c[0].lower() == 'isdeleted' for c in cols):
        out.append(' WHERE [IsDeleted] = 0;')
    else:
        out.append(';')
    out.append('\nEND\nGO\n')

    out.append(f'IF OBJECT_ID(N''Delete{table}ById'', N''P'') IS NOT NULL DROP PROCEDURE Delete{table}ById;\nGO')
    out.append(f'CREATE PROCEDURE Delete{table}ById\n(\n    @{pk} {param_type(next(c[1] for c in cols if c[0] == pk))}\n)\nAS\nBEGIN\n    SET NOCOUNT ON;\n')
    if any(c[0].lower() == 'isdeleted' for c in cols):
        out.append(f'    UPDATE [dbo].[{table}] SET [IsDeleted] = 1 WHERE [{pk}] = @{pk};\n')
    else:
        out.append(f'    DELETE FROM [dbo].[{table}] WHERE [{pk}] = @{pk};\n')
    out.append('    SELECT @@ROWCOUNT AS RowsAffected;\nEND\nGO\n')

    out.append(f'IF OBJECT_ID(N''ActiveInActive{table}ById'', N''P'') IS NOT NULL DROP PROCEDURE ActiveInActive{table}ById;\nGO')
    if any(c[0].lower() == 'isactive' for c in cols):
        out.append(f'CREATE PROCEDURE ActiveInActive{table}ById\n(\n    @{pk} {param_type(next(c[1] for c in cols if c[0] == pk))},\n    @IsActive BIT\n)\nAS\nBEGIN\n    SET NOCOUNT ON;\n    UPDATE [dbo].[{table}] SET [IsActive] = @IsActive WHERE [{pk}] = @{pk};\n    SELECT @@ROWCOUNT AS RowsAffected;\nEND\nGO\n')
    else:
        out.append(f'CREATE PROCEDURE ActiveInActive{table}ById\n(\n    @{pk} {param_type(next(c[1] for c in cols if c[0] == pk))},\n    @IsActive BIT\n)\nAS\nBEGIN\n    SET NOCOUNT ON;\n    RAISERROR(''Table [{table}] does not contain an IsActive column.'', 16, 1);\n    RETURN;\nEND\nGO\n')

output = '\n'.join(out)
output_path = Path(__file__).parent / 'Gayatri-Portal-DB-sprocs.sql'
output_path.write_text(output, encoding='utf-8')
print(f'Wrote {output_path.resolve()} with {len(tables)} tables.')
